Profile: add mincount to printout

vtjnash opened the pull request:
after normalizing lots of code in profile to make sure I didn't have to thread new properties manually through each function, I added a `mincount` cut off to `Profile.print` to make them shorter. I recommend something around 1-5% of the total number of samples. For example:

```
julia> Profile.print(mincount=200, maxdepth=50)
6198 ./event.jl:68; (::Base.REPL.##3#4{Base.REPL.REPLBackend})()
 6198 ./REPL.jl:97; macro expansion
  6197 ./REPL.jl:66; eval_user_input(::Any, ::Base.REPL.REPLBackend)
   6197 ./boot.jl:234; eval(::Module, ::Any)
    6197 ./<missing>:?; anonymous
     6197 ./profile.jl:16; macro expansion;
      6197 ./loading.jl:332; require(::Symbol)
       6152 ./loading.jl:202; _require_from_serialized(::Int64, ::Symbol, ::Bool)
        2621 ./loading.jl:173; _require_from_serialized(::Int64, ::Symbol, ::String, ::Bool)
         2621 ./loading.jl:579; recompile_stale(::Symbol, ::String)
          2605 ./loading.jl:562; stale_cachefile(::String, ::String)
           2605 ./loading.jl:332; require(::Symbol)
            2605 ./loading.jl:202; _require_from_serialized(::Int64, ::Symbol, ::Bool)
             2359 ./loading.jl:174; _require_from_serialized(::Int64, ::Symbol, ::String, ::Bool)
              245 /Users/jameson/.julia/v0.6/PyCall/src/pyinit.jl:50; __init__()
               236 ./version.jl:93; VersionNumber(::SubString{String})
        3531 ./loading.jl:174; _require_from_serialized(::Int64, ::Symbol, ::String, ::Bool)
         462 /Users/jameson/.julia/v0.6/PyPlot/src/PyPlot.jl:230; __init__()
          256 /Users/jameson/.julia/v0.6/PyPlot/src/PyPlot.jl:178; find_backend(::PyCall.PyObject)
```

(if you look at above with C=true, you can confirm that most of the overhead (~70%) is from llvm 3.7 compile time, with jl_recache_types adding ~10%)

vtjnash committed:
cleanup stacktraces method signatures

ensure that we are using consistent types
and introduce ProfileFormat to contain all of the user options

vtjnash committed:
profile: add mincount argument

the min-count argument to Profile.print can be used to
filter out lines from the display that are insignificant

yuyichao commented:
What's the easiest way to specify the mincount as a ratio?

vtjnash commented:
`count(ip == C_NULL for ip in bt)`?

tkelman commented on base/profile.jl:
```diff
@@ -66,45 +66,76 @@ Clear any existing backtraces from the internal buffer.
+(`:filefuncline` sorts by the source line, whereas `:count`
 sorts in order of number of collected samples).
+`mincount` can also be used to limit the printout to only those
+lines with at least mincount occurances.
```
occurrences

yuyichao commented:
> `sum(ip == C_NULL for ip in bt)`?

What about a way to specify that to `Profile.print` without first dumping the full profile to count how many samples are collected.

timholy commented:
One proposal would be to automatically truncate anything with fewer than `sqrt(N)` counts, where `N` is the number of counts in the "parent." `sqrt(N)` is a measure of the sampling noise.

vtjnash commented:
ooo, that sounds fancy and more even more accurate for quickly picking out the key frames. what would you call that option?

timholy commented:
That's a good question. `significant`? `prominent`?

timholy commented:
And to clarify, by `parent` I meant "immediate parent". If there were 10^4 samples taken at a particular line, one of its children with 101 samples should be displayed; of its children, anything with >11 samples should be shown. Maybe `n>ceil(sqrt(N))`? Once you get down to 3 samples, you have to retain all of them to be displayed.

JeffBezanson commented:
:+1: to this. Very good idea.

vtjnash commented:
shouldn't that be `n > floor(sqrt(N))`?

what do you think of `noisefloor`, used as a multiplier `n > floor(noisefloor * sqrt(N))` (with a default of 0, and suggested value of 1)?

timholy commented:
> shouldn't that be `n > floor(sqrt(N))`

only differs by 1 from what I wrote, so it will only matter for small numbers of samples. I don't care which you choose (there's no "in principle correct choice," in much the same way that the choice of a p-value is actually arbitrary),

> what do you think of `noisefloor`

:+1: . That's a great suggestion.

vtjnash committed:
profile: add noisefloor argument

the noise-floor argument to Profile.print can be used to
filter out lines from the display that are insignificant,
based on the heuristic `n <= noisefloor * sqrt(N)`

tkelman commented on base/profile.jl:
```diff
@@ -66,45 +66,92 @@ Clear any existing backtraces from the internal buffer.
-sorts in order of number of collected samples).
+will be used.
+
+The keywords arguments can be any combinations of:
```
keyword arguments

tkelman commented on base/profile.jl:
```diff
@@ -66,45 +66,92 @@ Clear any existing backtraces from the internal buffer.
+
+The keywords arguments can be any combinations of:
+
+ - `format` can be `:tree` or `:flat`.
```
all of these should specify what the default behavior is

tkelman commented on base/profile.jl:
```diff
@@ -66,45 +66,92 @@ Clear any existing backtraces from the internal buffer.
+   (`:filefuncline` sorts by the source line, whereas `:count`
+   sorts in order of number of collected samples).
+
+ - `noisefloor` only shows frames that exceed the heuristic noise floor of the sample (only applies to format :tree).
```
code highlight `:tree`
